arch/x86/core/userspace.S

```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright (c) 2017 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <kernel_structs.h>
+#include <arch/x86/asm.h>
+#include <offsets_short.h>
+
+GTEXT(_x86_syscall_invoke)
+GTEXT(_x86_syscall_entry_stub)
+GTEXT(_x86_syscall_entry)
+GTEXT(_x86_userspace_enter)
+
+/* Landing site for syscall SW IRQ. Marshal arguments and call C function for
+ * further processing.
+ */
+SECTION_FUNC(TEXT, _x86_syscall_entry_stub)
+	push	%esi		/* arg5 */
+	push	%edi		/* arg4 */
+	push	%ebx		/* arg3 */
+#ifndef CONFIG_X86_IAMCU
+	push	%ecx		/* arg2 */
+	push	%edx		/* arg1	*/
+	push	%eax		/* Call id */
+#endif
+
+	call	_x86_syscall_entry
+
+	/* EAX now contains return value. Pop or xor everything else to prevent
+	 * information leak from kernel mode.
```

Comment:
Started typing a comment here that popping the registers was likely a clumsy way to do this, but never mind, I'm wrong.  These are all 1-byte instructions (and they save the ESP adjustment too), and the extra memory work is essentially free on iamcu, and even on bigcore it's going to be cache-resident anyway.  Looks weird, but really clever.  Cool.